### PR TITLE
[18.0][MIG] mail_composer_cc_bcc: Migration to 18.0

### PR DESCRIPTION
**Pull request by trisdoan**
### Note
- Missing commits ported in https://github.com/trisdoan/social/pull/28

### Changes in 18.0
- uid was forced in https://github.com/odoo/odoo/commit/9bc32a3e3cd7ee792a43451fa4458126062aff93, which made some keys are missing compared to standard '_notify_get_recipients'

### This changes
[[IMP] mail_composer_cc_bcc: refactor _generate_template_recipients](https://github.com/trisdoan/social/pull/29/commits/3695b7781c32b37b46b6cf0cec724ff18b7dcd8b)
- Took this chance to refactor this long method (If we all agree, this commit will be backported to former version :wink:)
    - It was completely overridden in previous migration, which is not necessary
    - Also, it results in `Invalid field 'partner_bcc_ids' on model 'mail.mail` in case of mass mailing
- Briefly, the method looks for email_bcc/email_cc configured in mail template, then populates respective data in mail composer: `partner_bcc_ids`, `partner_cc_ids`

In standard, email_cc is stripped off later combined into partner_ids, as following
standard

With this module, I believe we are trying to mimic normal email form: 
- main recipients in "To"
- cc recipients in "Cc" and so on. 

=> That's why reprocessing email_cc is needed

cc_bcc_composer

### CODE REVIEW

**Comment by trisdoan** [comment]
~~DRAFT: inspect `_generate_template_recipients`~~

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]
- Tests failed
image


- Error:
 ```
  File "/home/chien/code/odoo/odoo/master/addons/mail/wizard/mail_compose_message.py", line 686, in action_send_mail
    self._action_send_mail(auto_commit=False)
  File "/home/chien/code/odoo/odoo/master/addons/mail/wizard/mail_compose_message.py", line 714, in _action_send_mail
    result_mails_su += wizard._action_send_mail_mass_mail(res_ids, auto_commit=auto_commit)
  File "/home/chien/code/odoo/odoo/master/addons/mail/wizard/mail_compose_message.py", line 763, in _action_send_mail_mass_mail
    iter_mails_sudo = self.env['mail.mail'].sudo().create(res_ids_values)
  File "<decorator-gen-139>", line 2, in create
  File "/home/chien/code/odoo/odoo/master/odoo/api.py", line 480, in _model_create_multi
    return create(self, arg)
  File "/home/chien/code/odoo/odoo/master/addons/mail/models/mail_mail.py", line 154, in create
    new_mails = super(MailMail, self).create(values_list)
  File "<decorator-gen-31>", line 2, in create
  File "/home/chien/code/odoo/odoo/master/odoo/api.py", line 480, in _model_create_multi
    return create(self, arg)
  File "/home/chien/code/odoo/odoo/master/odoo/models.py", line 4920, in create
    raise ValueError("Invalid field %r on model %r" % (key, self._name))
ValueError: Invalid field 'partner_bcc_ids' on model 'mail.mail'
```
Steps to reproduce:
1. Create a mail template (for `Contact` for example): input values for CC and BCC
2. click `Add Context Action` on that template form.
3. Navigate to the menu of `Contact`, open 1 record and click `Action > Send Mail (...)`
4. Click on `Send`

**Comment by trisdoan** [comment]
DRAFT: working on errors

**Comment by trisdoan** [comment]
Hi anh @nguyenminhchien, I just included fixes for error you found

Briefly,
1. There was an error: List index out of range, which was due to improper test record set-up. It's fixed as following (+ included in 17.0, in [here](https://github.com/trisdoan/social/pull/27)):
```
 def test_email_cc_bcc(self):
        self.test_record.email = "test@example.com"
```
3. For "Invalid field 'partner_bcc_ids' on model 'mail.mail'", I added a check to not process cc, bcc for mass mailing. It's fixed as following:

```
context = self.env.context
if context.get("default_composition_mode", "") != "comment":
  return res
```

File: mail_composer_cc_bcc/models/mail_template.py [hunk 1]
```diff
@@ -60,142 +47,47 @@ def _generate_template_recipients(  # noqa:C901
...
-            if partner_to:
-                tpl_partner_ids = (
-                    set(self._parse_partner_to(partner_to)) & existing_pids
-                )
-                record_values.setdefault("partner_ids", []).extend(tpl_partner_ids)
+                    for field in ["bcc", "cc"]:
+                        if original_email in email_to_res_ids[field]:
+                            for res_id in email_to_res_ids[field][original_email]:
+                                res[res_id].setdefault(
+                                    f"partner_{field}_ids", []
```
Comment (nguyenminhchien) [inline mail_composer_cc_bcc/models/mail_template.py]: returning the value of `partner_bcc_ids` and `partner_cc_ids` here is dangerous. We can not trust on the context ` context.get("default_composition_mode")`. 
The most important thing is that, all the key in returned value must be in the `render_fields`.
So, don't compute  `partner_bcc_ids` and `partner_cc_ids`  here. But, do this in `def _compute_partner_cc_bcc_ids(self):` of `mail.compose.message`.
```
                if rendered_values.get("email_cc"):
                    partners = self.env["res.partner"]._find_or_create_from_emails(.... from rendered_values.get("email_cc")....)
                    if partners:
                        partner_cc_ids = partners.ids
               ........ same for email_bcc ............................
```

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]
Tests failed
```
2024-11-05 08:38:21,312 162331 INFO v18c_mail_composer_cc_bcc odoo.addons.mail_composer_cc_bcc.tests.test_mail_cc_bcc: Starting TestMailCcBcc.test_template_cc_bcc ... 
2024-11-05 08:38:21,443 162331 INFO v18c_mail_composer_cc_bcc odoo.addons.mail_composer_cc_bcc.tests.test_mail_cc_bcc: ====================================================================== 
2024-11-05 08:38:21,443 162331 ERROR v18c_mail_composer_cc_bcc odoo.addons.mail_composer_cc_bcc.tests.test_mail_cc_bcc: FAIL: TestMailCcBcc.test_template_cc_bcc
Traceback (most recent call last):
  File "/home/chien/code/oca/social/18.0/mail_composer_cc_bcc/tests/test_mail_cc_bcc.py", line 151, in test_template_cc_bcc
    self.assertEqual(composer.partner_cc_ids, expecting)
AssertionError: res.partner(40,) != res.partner(40, 7)
 
```

File: mail_composer_cc_bcc/models/mail_template.py (hunk 1 above)
Comment (trisdoan) [reply mail_composer_cc_bcc/models/mail_template.py]: Hi anh @nguyenminhchien, indeed, "context" looks like not a good approach

I just have an idea, how does it sound to you ?
- As we decided to not interfere with mass_mailing, in `https://github.com/OCA/social/pull/1203`
- So, patch `_prepare_mail_values` to filter out add-on fields

Pros
- Keep consistency with Odoo's flow: recipients are generated in `_generate_template_recipients`
- Processing those mails once. If doing that in `_compute_partner_cc_bcc_ids`, changing any dependency results in re-trigger.

Cons
- Yes, quite resource-consuming (adding cc/bcc mail back and forth). But is it common to have cc/bcc in context of mass mailing :thinking:.

Comment (nguyenminhchien) [reply mail_composer_cc_bcc/models/mail_template.py]: Hi Trí,
You should not return the value of any field which is not in `render_fields` (method `def _generate_template_recipients()`). Particularly,  `partner_cc_ids` and `partner_bcc_ids` are not in `render_fields`. So, you should not include them in the returned value.
image

Comment (nguyenminhchien) [reply mail_composer_cc_bcc/models/mail_template.py]: I think `def _compute_partner_cc_bcc_ids` is the best suited-place where to handle computing the value of `partner_cc_ids` and `partner_bcc_ids`.

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]

**Review by nguyenminhchien — APPROVED** [review]

**Pull request closed without merging**